File: src/lib/quant/costs.ts
/**
 * What a round turn actually costs, decomposed into the things a broker statement itemises.
 *
 * The stack this replaces had ONE number, `commissionRoundTurn`, standing in for four different
 * charges plus a flat tick of slippage. That is fine for an order-of-magnitude sanity check and
 * wrong for everything else, because the components behave differently:
 *
 *   * BROKER COMMISSION is per side, negotiable, and the only part that varies much between
 *     retail accounts. It is also the part people quote when they say "commissions".
 *   * EXCHANGE + CLEARING FEES are per side and set by CME, not the broker. They are the reason a
 *     micro contract is not a tenth of the cost of an E-mini: MNQ carries a tenth of the tick
 *     value against roughly a third of the fee, so the cost per point traded is ~3x worse.
 *   * REGULATORY (NFA) is per side and tiny, but it is per side, so it scales with turnover the
 *     same way everything else does.
 *   * SPREAD is not a fee at all. It is paid only when you TAKE liquidity, so an exit at a resting
 *     limit does not pay it and a market stop-out pays it twice over.
 *   * SLIPPAGE is not a constant. It is the part that gets worse exactly when a strategy needs it
 *     not to — a stop triggering in a fast market is a market order into a thin book.
 *
 * Separating them matters because they scale differently with the thing a study varies. Doubling
 * trade frequency doubles all of it; halving the stop distance leaves the dollar cost unchanged
 * while halving the edge it is measured against; moving from NQ to MNQ cuts tick value by 10 and
 * fees by ~3.
 *
 * ON THE NUMBERS BELOW. The structure here is exact. The VALUES are assumptions, dated, and must
 * be replaced with your own broker's statement and the current CME fee schedule before any of this
 * is used to size real risk. Exchange fees change, they differ by membership tier and by whether
 * the trade is electronic, and broker commissions differ by volume tier and by account. Nothing in
 * this file should be treated as a quote. `costSensitivity()` in `robustness.ts` exists precisely
 * because these are guesses: it reports how much of a result survives if they are wrong by 2x, and
 * that is the number to trust, not the point estimate.
 */
import type { Instrument } from "./types";

/** Per-side charges, in USD per contract. Every one of these appears on a real statement. */
export interface FeeSchedule {
  /** Broker's own commission, per side, per contract. The negotiable part. */
  brokerPerSide: number;
  /** CME exchange fee, per side. Set by the exchange; varies by product and membership tier. */
  exchangePerSide: number;
  /** Clearing fee, per side. Often bundled into the exchange line on a statement. */
  clearingPerSide: number;
  /** NFA regulatory fee, per side. */
  regulatoryPerSide: number;
  /** Free-text note on where these came from and what tier they assume. */
  source: string;
}

export const NO_FEES: FeeSchedule = {
  brokerPerSide: 0,
  exchangePerSide: 0,
  clearingPerSide: 0,
  regulatoryPerSide: 0,
  source: "zero — for isolating gross edge, never for a tradeable result",
};

export function feesPerSide(f: FeeSchedule): number {
  return f.brokerPerSide + f.exchangePerSide + f.clearingPerSide + f.regulatoryPerSide;
}

export function feesRoundTurn(f: FeeSchedule): number {
  return 2 * feesPerSide(f);
}

/**
 * Slippage model. A flat tick is the assumption most backtests make and the one that flatters a
 * stop-loss strategy most, because the flat tick is charged in the calm bars too while the real
 * cost concentrates in the fast ones.
 *
 * `base` is the quiet-market cost in ticks. `volCoef` scales it by how fast this bar is relative to
 * the instrument's own typical bar: a bar at 3x its median true range charges
 * `base * (1 + volCoef * (3 - 1))`. `stopExtra` is charged ON TOP for a stop exit, because a stop
 * becomes a market order at the worst moment by construction — the book is thinnest and moving
 * away. `illiquidMult` multiplies everything outside the instrument's own session.
 *
 * This is a MODEL, not a measurement. Bars are not order books and nothing here can be calibrated
 * from OHLCV alone. What it buys is that the cost stops being independent of the conditions the
 * strategy trades in, which is the specific way a flat tick lies.
 */
export interface SlippageModel {
  base: number;
  volCoef: number;
  stopExtra: number;
  illiquidMult: number;
  /**
   * Ceiling on the volatility stretch. Without it one freak bar -- a limit move, a bad print that
   * survived the audit -- sets the cost of the whole study, and a model that can be dominated by
   * its own tail is not measuring the thing it claims to. Real slippage does blow out in a fast
   * market, but it blows out into a book that still exists.
   */
  maxStretch: number;
}

export const FLAT_SLIPPAGE: SlippageModel = { base: 1, volCoef: 0, stopExtra: 0, illiquidMult: 1, maxStretch: 1 };

/** The default: slippage that grows with the bar's own speed and doubles outside the session. */
/**
 * Calibrated so a CALM, in-session, market-in / market-out round turn costs what the old flat
 * model charged — 1 tick of spread plus 1 tick of slippage per side. That is deliberate: a
 * realism change must not quietly hand back a discount, or every result that improves is
 * unattributable. From that baseline it only ever gets WORSE: a stop pays `stopExtra` on top
 * because it is a market order into a book that is moving away, a bar running at 3x its median
 * true range pays double, and a fill outside the session pays double again.
 */
export const REALISTIC_SLIPPAGE: SlippageModel = { base: 1, volCoef: 0.5, stopExtra: 1, illiquidMult: 2, maxStretch: 3 };

/**
 * What kind of fill this was, which is what decides whether the spread and slippage are paid.
 *
 *   taker  crossed the spread to get filled — a market entry, a time or session exit, and a target
 *          exit under the pessimistic `taker` fill model.
 *   stop   a protective stop that triggered. A taker fill by construction, at the worst moment:
 *          the book is thinnest and moving away, so it carries an extra slippage term.
 *   maker  a resting order that price came to — a limit entry, or a target under a fill model that
 *          lets targets rest. Pays no spread and no slippage, but is not guaranteed to fill at all,
 *          which the engines model separately by requiring price to trade through the level.
 */
export type OrderRole = "taker" | "stop" | "maker";

export interface FillContext {
  /** This bar's true range over the instrument's median true range. 1 = a typical bar. */
  volRatio: number;
  /** False when the fill lands outside the instrument's own session. */
  inSession: boolean;
}

export const CALM: FillContext = { volRatio: 1, inSession: true };

/** Slippage in TICKS for one fill, given what kind of order it was and what the bar looked like. */
export function slippageTicks(m: SlippageModel, role: OrderRole, ctx: FillContext = CALM): number {
  if (role === "maker") return 0; // a resting limit is hit; it does not chase
  const vol = Math.max(ctx.volRatio, 0);
  const stretch = Math.min(1 + m.volCoef * Math.max(vol - 1, 0), Math.max(m.maxStretch, 1));
  let t = m.base * stretch;
  if (role === "stop") t += m.stopExtra * stretch;
  if (!ctx.inSession) t *= m.illiquidMult;
  return t;
}

/**
 * The complete cost of one fill, in PRICE units.
 *
 * A maker exit pays neither spread nor slippage — it was already resting when price came to it.
 * Everything else crosses half the spread and pays the modelled slippage. Fees are per side and
 * charged on every fill regardless of how it was filled, because the exchange does not care.
 */
export function fillCostPoints(inst: Instrument, role: OrderRole, ctx: FillContext = CALM): number {
  const fees = effectiveFees(inst);
  const slip = effectiveSlippage(inst);
  const perSideUsd = feesPerSide(fees);
  const usdPerPoint = inst.tickValue / inst.tickSize;
  const feePoints = perSideUsd / usdPerPoint;
  if (role === "maker") return feePoints;
  const spreadPoints = (inst.spreadTicks / 2) * inst.tickSize;
  return feePoints + spreadPoints + slippageTicks(slip, role, ctx) * inst.tickSize;
}

/**
 * A whole round turn under the given fill roles, in price units — the number that decides whether
 * a scalp can exist at all. Default is the pessimistic case: market in, stopped out.
 */
/**
 * The spread and slippage half of a fill, in PRICE units — everything EXCEPT the per-side fees.
 *
 * Split out because the two halves have different shelf lives. Fees are a constant per trade, so
 * they can be applied when a result is read, which keeps trying a different broker free. Spread
 * and slippage depend on the bar the fill landed on, so they must be computed during the price
 * walk. `tuner/tensor.ts` caches this part per trade and adds the fees when the result is read.
 */
export function fillFrictionPoints(inst: Instrument, role: OrderRole, ctx: FillContext = CALM): number {
  if (role === "maker") return 0;
  const slip = effectiveSlippage(inst);
  return (inst.spreadTicks / 2) * inst.tickSize + slippageTicks(slip, role, ctx) * inst.tickSize;
}

/** Per-side fees expressed in price units. Constant per fill, so it is applied at read time. */
/**
 * Which slippage model actually applies, resolving the same disagreement `effectiveFees` does.
 *
 * `slippageTicks` is the headline and `slippage.base` is the detail behind it, normally equal. A
 * study or test may override the headline alone -- `{ ...inst, slippageTicks: 0 }` to isolate
 * gross edge -- and if the model simply won, that override would silently do nothing. So when the
 * two disagree the SCALAR wins, applied flat: an explicit override is a statement of intent.
 */
export function effectiveSlippage(inst: Instrument): SlippageModel {
  const m = inst.slippage;
  if (m && Math.abs(m.base - inst.slippageTicks) < 1e-9) return m;
  return { base: inst.slippageTicks, volCoef: 0, stopExtra: 0, illiquidMult: 1, maxStretch: 1 };
}

export function feePoints(inst: Instrument): number {
  return feesPerSide(effectiveFees(inst)) / (inst.tickValue / inst.tickSize);
}

export function roundTurnPoints(
  inst: Instrument,
  entry: OrderRole = "taker",
  exit: OrderRole = "stop",
  ctx: FillContext = CALM,
): number {
  return fillCostPoints(inst, entry, ctx) + fillCostPoints(inst, exit, ctx);
}

/**
 * Back-compat shim: an instrument declared with only the old lumped `commissionRoundTurn` is read
 * as broker commission split evenly across the two sides, with no exchange or regulatory line.
 * That UNDERSTATES the true cost, which is the honest direction for a shim to be wrong in — it
 * keeps an un-migrated instrument's numbers identical to what they were rather than silently
 * improving them, so a change in a result is always traceable to a deliberate edit.
 */
/**
 * Which fee block actually applies, resolving the one way these two fields can disagree.
 *
 * `fees` is the detail and `commissionRoundTurn` is the headline, and normally the second is
 * derived from the first. But a study or a test may override the headline alone -- `{ ...inst,
 * commissionRoundTurn: 0 }` to isolate gross edge is a real and reasonable thing to write -- and
 * if `fees` simply won, that override would silently do nothing and the study would report a
 * costed result while believing it had none. So: when the two disagree, the LUMPED NUMBER WINS,
 * because an explicit override is a statement of intent and the derived detail is not.
 */
export function effectiveFees(inst: Instrument): FeeSchedule {
  const f = inst.fees;
  if (f && Math.abs(feesRoundTurn(f) - inst.commissionRoundTurn) < 1e-9) return f;
  return legacyFees(inst);
}

export function legacyFees(inst: Instrument): FeeSchedule {
  return {
    brokerPerSide: inst.commissionRoundTurn / 2,
    exchangePerSide: 0,
    clearingPerSide: 0,
    regulatoryPerSide: 0,
    source: "legacy lumped commissionRoundTurn; no exchange or regulatory line",
  };
}

// ---------------------------------------------------------------------------- broker presets
/**
 * Retail futures cost profiles, as component breakdowns rather than a single number.
 *
 * READ THIS BEFORE USING ANY OF THEM. These are dated assumptions at roughly the right magnitude
 * for a small non-member retail account, not quotes. Broker commissions vary by volume tier and by
 * negotiation; exchange and clearing fees are set by CME and change; membership tiers change them
 * a lot. Take the numbers off your own statement — every field is separately settable and
 * `describe()` prints the breakdown so a mismatch is obvious.
 */
export interface BrokerPreset {
  id: string;
  label: string;
  /** Broker commission per side, per contract, by contract class. */
  brokerPerSide: { micro: number; emini: number };
  note: string;
}

export const BROKER_PRESETS: Record<string, BrokerPreset> = {
  discount: {
    id: "discount",
    label: "Discount futures broker (Tradovate / NinjaTrader / AMP tier)",
    brokerPerSide: { micro: 0.35, emini: 0.85 },
    note: "Typical unbundled retail rate; lifetime-licence and high-volume tiers go lower.",
  },
  ibkr: {
    id: "ibkr",
    label: "Interactive Brokers, tiered",
    brokerPerSide: { micro: 0.25, emini: 0.85 },
    note: "Tiered schedule; IBKR's fixed tier bundles exchange fees into a single larger number.",
  },
  premium: {
    id: "premium",
    label: "Full-service / low-volume account",
    brokerPerSide: { micro: 0.75, emini: 2.25 },
    note: "What a small account without a negotiated rate typically pays.",
  },
  propfirm: {
    id: "propfirm",
    label: "Prop-firm evaluation account",
    brokerPerSide: { micro: 0.5, emini: 1.5 },
    note: "Evaluation accounts usually charge a flat all-in rate; check whether it already includes exchange fees.",
  },
};

/** CME per-side exchange + clearing, by product. Assumption, non-member electronic. */
export const EXCHANGE_FEES: Record<string, { exchange: number; clearing: number; class: "micro" | "emini" }> = {
  NQ: { exchange: 1.18, clearing: 0.0, class: "emini" },
  ES: { exchange: 1.18, clearing: 0.0, class: "emini" },
  MNQ: { exchange: 0.35, clearing: 0.0, class: "micro" },
  MES: { exchange: 0.35, clearing: 0.0, class: "micro" },
  GC: { exchange: 1.55, clearing: 0.0, class: "emini" },
  MGC: { exchange: 0.55, clearing: 0.0, class: "micro" },
  CL: { exchange: 1.55, clearing: 0.0, class: "emini" },
  MCL: { exchange: 0.55, clearing: 0.0, class: "micro" },
};

/** NFA regulatory fee per side, per contract. */
export const NFA_PER_SIDE = 0.02;

/** As-of date for every fee number in this file. Bump it when you re-check the schedules. */
export const FEES_AS_OF = "2026-08 — ASSUMPTION, not a quote. Verify against your statement.";

export function scheduleFor(instrumentId: string, brokerId = "discount"): FeeSchedule {
  const ex = EXCHANGE_FEES[instrumentId.toUpperCase()];
  const broker = BROKER_PRESETS[brokerId];
  if (!broker) throw new Error(`unknown broker preset "${brokerId}" (known: ${Object.keys(BROKER_PRESETS).join(", ")})`);
  if (!ex) {
    return {
      brokerPerSide: broker.brokerPerSide.emini,
      exchangePerSide: 0,
      clearingPerSide: 0,
      regulatoryPerSide: NFA_PER_SIDE,
      source: `${broker.label}; no exchange schedule known for ${instrumentId} — exchange fee NOT included, so this UNDERSTATES cost`,
    };
  }
  return {
    brokerPerSide: broker.brokerPerSide[ex.class],
    exchangePerSide: ex.exchange,
    clearingPerSide: ex.clearing,
    regulatoryPerSide: NFA_PER_SIDE,
    source: `${broker.label}, ${ex.class} class, ${FEES_AS_OF}`,
  };
}

/** A human-readable breakdown — the thing to hold next to a statement to check it. */
export function describe(inst: Instrument): string {
  const f = effectiveFees(inst);
  const usdPerPoint = inst.tickValue / inst.tickSize;
  const rt = roundTurnPoints(inst);
  const lines = [
    `${inst.id} — ${inst.label}`,
    `  tick ${inst.tickSize} = $${inst.tickValue.toFixed(2)}; 1 point = $${usdPerPoint.toFixed(2)}`,
    `  broker      $${f.brokerPerSide.toFixed(2)}/side`,
    `  exchange    $${f.exchangePerSide.toFixed(2)}/side`,
    `  clearing    $${f.clearingPerSide.toFixed(2)}/side`,
    `  regulatory  $${f.regulatoryPerSide.toFixed(2)}/side`,
    `  fees        $${feesRoundTurn(f).toFixed(2)} round turn`,
    `  spread      ${inst.spreadTicks} tick(s) crossed once = $${(inst.spreadTicks * inst.tickValue).toFixed(2)}`,
    `  slippage    ${JSON.stringify(effectiveSlippage(inst))}`,
    `  ROUND TURN  ${(rt / inst.tickSize).toFixed(2)} ticks = $${(rt * usdPerPoint).toFixed(2)}  (market in, stopped out, calm bar)`,
    `  source      ${f.source}`,
  ];
  return lines.join("\n");
}

File: research/costs.py
"""What a round turn actually costs, itemised the way a broker statement itemises it.

The research layer has carried `PV = 2.0; TICK = 0.25; COMM = 1.0; EC = 2*TICK; SE = 1*TICK` as
bare module constants, copy-pasted into about twenty files. Two problems with that, and the second
is the expensive one:

  * COMM = 1.00 per round turn is BROKER COMMISSION ONLY. It has no CME exchange fee and no NFA
    line. On MNQ the exchange fee is about a third of the broker's own charge again, so the real
    round turn is nearer $1.44 than $1.00 -- every result in `docs/ib/` was measured about 44%
    light on fees.
  * EC and SE are flat. A flat tick is charged in the calm bars where it is not paid and
    understated in the fast ones where it is, and a stop-loss strategy is exactly the thing that
    exits preferentially into fast bars. That is not a small bias and it is not symmetric.

This module is the single definition, mirroring `src/lib/quant/costs.ts` so the two engines cannot
drift. The STRUCTURE here is exact. The VALUES are dated assumptions -- exchange fees change,
membership tiers change them a lot, and broker commissions vary by volume -- so replace them with
your own statement before sizing real risk. `mult` exists to sweep exactly that uncertainty.
"""
from __future__ import annotations

from dataclasses import dataclass, replace

import numpy as np

TICK = 0.25
NFA_PER_SIDE = 0.02
FEES_AS_OF = "2026-08 - ASSUMPTION, not a quote. Verify against your statement."

# CME per-side exchange + clearing, non-member electronic. Assumption.
EXCHANGE = {
    "NQ": (1.18, 0.0, "emini"), "ES": (1.18, 0.0, "emini"),
    "MNQ": (0.35, 0.0, "micro"), "MES": (0.35, 0.0, "micro"),
    "GC": (1.55, 0.0, "emini"), "MGC": (0.55, 0.0, "micro"),
    "CL": (1.55, 0.0, "emini"), "MCL": (0.55, 0.0, "micro"),
}

# Broker commission per side, per contract, by contract class.
BROKERS = {
    "discount": dict(micro=0.35, emini=0.85,
                     label="Discount futures broker (Tradovate / NinjaTrader / AMP tier)"),
    "ibkr": dict(micro=0.25, emini=0.85, label="Interactive Brokers, tiered"),
    "premium": dict(micro=0.75, emini=2.25, label="Full-service / low-volume account"),
    "propfirm": dict(micro=0.50, emini=1.50, label="Prop-firm evaluation account"),
    "legacy": dict(micro=0.50, emini=2.00,
                   label="the old COMM=1.00 round turn, broker only, no exchange or NFA line"),
}


@dataclass(frozen=True)
class Fees:
    """Per side, per contract, in USD. Every line appears on a real statement."""
    broker: float
    exchange: float
    clearing: float
    regulatory: float
    source: str = ""

    def per_side(self):
        return self.broker + self.exchange + self.clearing + self.regulatory

    def round_turn(self):
        return 2.0 * self.per_side()


@dataclass(frozen=True)
class Slippage:
    """Slippage in TICKS, as a function of the bar a fill landed on and the role it played.

    A stop pays `stop_extra` on top of the taker cost because it is a market order into a book
    that is moving away, a fast bar pays more, and a fill outside the session pays double.

    Note the calm-bar case is NOT a match for what `sim_core` charged: the old model billed a flat
    2 ticks per side, this bills half a 1-tick spread plus `base`. On MNQ that is 1.5 ticks, so a
    strategy exiting on targets in quiet markets can come out slightly cheaper while one exiting
    on stops in fast markets comes out much worse. `research/real_costs.py` prints the fee and
    friction lines separately for exactly that reason -- the net direction is a property of a
    strategy's exit mix, not something the model decides in advance.
    """
    base: float = 1.0
    vol_coef: float = 0.5
    stop_extra: float = 1.0
    illiquid_mult: float = 2.0
    max_stretch: float = 3.0     # one freak bar must not set the cost of a whole study


FLAT = Slippage(base=1.0, vol_coef=0.0, stop_extra=0.0, illiquid_mult=1.0, max_stretch=1.0)
REALISTIC = Slippage()

# The model the research layer actually used before this change, reconstructed exactly so a
# before/after comparison never has to be remembered. `sim_core` charged EC = 2 ticks of spread
# plus slippage on EACH side and SE = 1 extra tick on a stop -- so the old model was never quite
# flat, it already had a stop premium; what it lacked was any dependence on the bar.
LEGACY_SLIP = Slippage(base=1.0, vol_coef=0.0, stop_extra=1.0, illiquid_mult=1.0, max_stretch=1.0)
LEGACY_SPREAD_TICKS = 2.0

STOP, TARGET = 1, 2


@dataclass(frozen=True)
class Costs:
    """The complete cost model for one instrument."""
    symbol: str = "MNQ"
    pv: float = 2.0                 # USD per point
    tick: float = TICK
    fees: Fees = None
    slip: Slippage = REALISTIC
    spread_ticks: float = 1.0
    # "taker" charges the spread on every fill including a target; "realistic" lets a target rest.
    fill_model: str = "taker"
    mult: float = 1.0               # scales everything; 2 is the standard stress case

    def with_broker(self, broker):
        return replace(self, fees=schedule(self.symbol, broker))

    def fee_points(self):
        return self.fees.per_side() / self.pv * self.mult

    def friction_ticks(self, role, vol_ratio=1.0, in_session=True):
        """Spread + slippage for one fill, in ticks. A maker fill pays neither."""
        if role == "maker":
            return 0.0
        m = self.slip
        stretch = min(1.0 + m.vol_coef * max(vol_ratio - 1.0, 0.0), max(m.max_stretch, 1.0))
        t = m.base * stretch
        if role == "stop":
            t += m.stop_extra * stretch
        if not in_session:
            t *= m.illiquid_mult
        return (self.spread_ticks / 2.0 + t)

    def friction_points(self, role, vol_ratio=1.0, in_session=True):
        return self.friction_ticks(role, vol_ratio, in_session) * self.tick * self.mult

    def round_turn_points(self, entry="taker", exit="stop"):
        return (2 * self.fee_points()
                + self.friction_points(entry) + self.friction_points(exit))

    def describe(self):
        f = self.fees
        rt_calm = 2 * self.fee_points() + 2 * self.friction_points("taker")
        rt_stop = self.round_turn_points()
        return "\n".join([
            f"{self.symbol}   1 point = ${self.pv:.2f}, tick {self.tick} = ${self.pv*self.tick:.2f}",
            f"  broker      ${f.broker:.2f}/side",
            f"  exchange    ${f.exchange:.2f}/side",
            f"  clearing    ${f.clearing:.2f}/side",
            f"  regulatory  ${f.regulatory:.2f}/side",
            f"  fees        ${f.round_turn():.2f} round turn",
            f"  spread      {self.spread_ticks:g} tick crossed once",
            f"  slippage    base {self.slip.base:g}t, vol_coef {self.slip.vol_coef:g}, "
            f"stop +{self.slip.stop_extra:g}t, illiquid x{self.slip.illiquid_mult:g}, "
            f"cap x{self.slip.max_stretch:g}",
            f"  ROUND TURN  {rt_calm/self.tick:.2f} ticks = ${rt_calm*self.pv:.2f}  "
            f"(market in, market out, calm bar)",
            f"              {rt_stop/self.tick:.2f} ticks = ${rt_stop*self.pv:.2f}  "
            f"(market in, STOPPED out, calm bar)",
            f"  source      {f.source}",
        ])


def schedule(symbol="MNQ", broker="discount") -> Fees:
    sym = symbol.upper()
    b = BROKERS.get(broker)
    if b is None:
        raise KeyError(f"unknown broker {broker!r}; known: {', '.join(BROKERS)}")
    ex = EXCHANGE.get(sym)
    if ex is None:
        return Fees(b["emini"], 0.0, 0.0, NFA_PER_SIDE,
                    f"{b['label']}; no exchange schedule for {sym} - fee NOT included, "
                    f"so this UNDERSTATES cost")
    if broker == "legacy":
        # The old model on purpose: broker only, no exchange, no NFA. Kept so a before/after
        # comparison is possible without reconstructing it from memory.
        return Fees(b[ex[2]], 0.0, 0.0, 0.0, "LEGACY: the old COMM, broker only")
    return Fees(b[ex[2]], ex[0], ex[1], NFA_PER_SIDE,
                f"{b['label']}, {ex[2]} class, {FEES_AS_OF}")


def model(symbol="MNQ", broker="discount", **kw) -> Costs:
    pv = {"NQ": 20.0, "MNQ": 2.0, "ES": 50.0, "MES": 5.0}.get(symbol.upper(), 2.0)
    return Costs(symbol=symbol.upper(), pv=pv, fees=schedule(symbol, broker), **kw)


LEGACY = model("MNQ", "legacy", slip=LEGACY_SLIP, spread_ticks=LEGACY_SPREAD_TICKS)
DEFAULT = model("MNQ", "discount")


def vol_ratio(h, l, c):
    """Each bar's true range over the series MEDIAN true range.

    Median rather than mean: bar ranges are heavy-tailed and a mean is dragged up by exactly the
    fast bars the model is trying to charge extra for, flattening the effect being measured."""
    pc = np.r_[c[0], c[:-1]]
    tr = np.maximum.reduce([h - l, np.abs(h - pc), np.abs(l - pc)])
    med = np.median(tr[np.isfinite(tr) & (tr > 0)]) if np.isfinite(tr).any() else 0.0
    return np.where(med > 0, tr / max(med, 1e-12), 1.0)


def friction_arrays(cost: Costs, h, l, c, mod, session=(570, 960)):
    """Per-BAR spread+slippage for a taker fill and for a stop fill, in points.

    Per bar rather than per trade is what keeps the cost a lookup: friction depends only on the bar
    a fill landed on and the role it played, and both engines already know the exit bar and the
    exit reason. So the broker, the fill model and the cost multiplier all stay free to change
    without recomputing anything.
    """
    vr = vol_ratio(h, l, c)
    lo, hi = session
    ins = (mod >= lo) & (mod < hi) if lo <= hi else (mod >= lo) | (mod < hi)
    m = cost.slip
    stretch = np.minimum(1.0 + m.vol_coef * np.maximum(vr - 1.0, 0.0), max(m.max_stretch, 1.0))
    taker_t = m.base * stretch
    stop_t = taker_t + m.stop_extra * stretch
    out_mult = np.where(ins, 1.0, m.illiquid_mult)
    half = cost.spread_ticks / 2.0
    f_taker = (half + taker_t * out_mult) * cost.tick * cost.mult
    f_stop = (half + stop_t * out_mult) * cost.tick * cost.mult
    return np.ascontiguousarray(f_taker), np.ascontiguousarray(f_stop)


if __name__ == "__main__":
    for sym in ("MNQ", "NQ"):
        print(model(sym, "discount").describe(), "\n")
    print("broker presets, MNQ round-turn fees:")
    for b in BROKERS:
        print(f"  {b:<10} ${schedule('MNQ', b).round_turn():.2f}   {BROKERS[b]['label']}")
